api/client/src/main/java/org/projectnessie/client/auth/oauth2/AuthorizationCodeFlow.java

```diff
@@ -0,0 +1,258 @@
+    this.console = console;
+    this.flowTimeout = params.getAuthorizationCodeFlowTimeout();
+    authCodeFuture = requestFuture.thenApply(this::extractAuthorizationCode);
+    tokensFuture = authCodeFuture.thenApply(this::fetchNewTokens);
+    closeFuture.thenRun(this::doClose);
+    server =
+        createServer(params.getAuthorizationCodeFlowWebServerPort().orElse(-1), this::doRequest);
+    state = OAuth2Utils.randomAlphaNumString(STATE_LENGTH);
+    redirectUri =
+        String.format("http://localhost:%d%s", server.getAddress().getPort(), CONTEXT_PATH);
+    URI authEndpoint = params.getAuthEndpoint().orElseThrow(IllegalStateException::new);
+    authorizationUri =
+        new UriBuilder(authEndpoint.resolve("/"))
+            .path(authEndpoint.getPath())
+            .queryParam("response_type", "code")
+            .queryParam("client_id", params.getClientId())
+            .queryParam("scope", params.getScope().orElse(null))
+            .queryParam("redirect_uri", redirectUri)
+            .queryParam("state", state)
+            .build();
+    LOGGER.debug("Authorization Code Flow: started, redirect URI: {}", redirectUri);
+  }
+
+  @Override
+  public void close() {
+    closeFuture.complete(null);
+  }
+
+  private void doClose() {
+    inflightRequestsPhaser.arriveAndAwaitAdvance();
+    LOGGER.debug("Authorization Code Flow: closing");
+    server.stop(0);
+    // don't close the console, it's not ours
+  }
+
+  private void abort() {
+    tokensFuture.cancel(true);
+    authCodeFuture.cancel(true);
+    requestFuture.cancel(true);
+  }
+
+  public Tokens fetchNewTokens() {
+    console.println();
+    console.println(MSG_PREFIX + "======= Nessie authentication required =======");
+    console.println(MSG_PREFIX + "Browse to the following URL to continue:");
+    console.println(MSG_PREFIX + authorizationUri);
+    console.println();
+    console.flush();
+    try {
+      return tokensFuture.get(flowTimeout.toMillis(), TimeUnit.MILLISECONDS);
+    } catch (TimeoutException e) {
+      LOGGER.error(MSG_PREFIX + "Timed out waiting for authorization code.");
+      abort();
+      throw new RuntimeException(e);
+    } catch (InterruptedException e) {
+      abort();
+      Thread.currentThread().interrupt();
+      throw new RuntimeException(e);
+    } catch (ExecutionException e) {
+      abort();
+      Throwable cause = e.getCause();
+      LOGGER.error(MSG_PREFIX + "Authentication failed: " + cause.getMessage());
+      if (cause instanceof HttpClientException) {
+        throw (HttpClientException) cause;
+      }
+      throw new RuntimeException(cause);
+    }
+  }
+
+  private void doRequest(HttpExchange exchange) {
+    LOGGER.debug("Authorization Code Flow: received request");
+    inflightRequestsPhaser.register();
+    requestFuture.complete(exchange);
+    tokensFuture
+        .handle((tokens, error) -> doResponse(exchange, error))
+        .whenComplete((exc, error) -> exc.close())
```

Comment:
`exc` could be `null` in case of a (runtime) exception... Should we use the `exchange` method parameter instead?